### PR TITLE
feat: add ingresses/routes page (#4367)

### LINKED ISSUE #4367: Kubernetes Ingress / Routes page
### Is your feature request related to a problem? Please describe

The podify flow creates ingress or Routes on Kubernetes, but there is no way to view them or clean them up from within Podman Desktop, and hence rerunning the flow doesn't work.

### Describe the solution you'd like

Add a new main nav item for Kubernetes Ingress & Routes, a list page, and details page - similar to what we have for Containers or Pods. Users should be able to delete either.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by lstocchi**
### What does this PR do?

it adds the ingresses/routes page

### Screenshot / video of UI

in progress

### What issues does this PR fix or reference?

it closes #4367 

### How to test this PR?

in progress

### CODE REVIEW

File: packages/renderer/src/lib/ingresses-routes/IngressRouteColumnHostPath.svelte [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+</script>
+
+{#if 'rules' in object && object.rules}
+  {#each object.rules as rule}
+    {@const host = rule.host ?? ''}
+    {#if rule.http}
+      {#each rule.http.paths as path}
+        {#if path.path}
+          {#if host}
+            <div class="text-sm text-gray-300">
```
Comment (deboer-tim) [inline packages/renderer/src/lib/ingresses-routes/IngressRouteColumnHostPath.svelte]: Should be text-gray-500.

File: packages/renderer/src/lib/ingresses-routes/IngressRouteColumnHostPath.svelte [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+              <Link externalRef="https://{host}{path.path}">
+                {host}{path.path}
+              </Link>
+            </div>
+          {:else}
+            <div class="text-sm text-gray-300">{host}{path.path}</div>
```
Comment (deboer-tim) [inline packages/renderer/src/lib/ingresses-routes/IngressRouteColumnHostPath.svelte]: Should be text-gray-500.

File: packages/renderer/src/lib/ingresses-routes/IngressRouteColumnBackend.svelte [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+export let object: IngressUI | RouteUI;
+</script>
+
+{#if 'rules' in object && object.rules}
+  {#each object.rules as rule}
+    {#if rule.http}
+      {#each rule.http.paths as path}
+        {#if path.backend.service}
+          <div class="text-sm text-gray-300">{path.backend.service.name}:{path.backend.service.port?.number ?? ''}</div>
```
Comment (deboer-tim) [inline packages/renderer/src/lib/ingresses-routes/IngressRouteColumnBackend.svelte]: Should be text-gray-500

File: packages/renderer/src/lib/ingresses-routes/IngressRouteColumnBackend.svelte [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+        {:else if path.backend.resource}
+          <div class="text-sm text-gray-300">{path.backend.resource.kind} {path.backend.resource.name}</div>
```
Comment (deboer-tim) [inline packages/renderer/src/lib/ingresses-routes/IngressRouteColumnBackend.svelte]: Should be text-gray-500

File: packages/renderer/src/lib/table/Table.svelte [hunk 5]
```diff
@@ -165,7 +165,7 @@ function setGridColumns() {
   <div role="rowgroup">
     {#each data as object (object)}
       <div
-        class="grid grid-table gap-x-0.5 mx-5 h-12 bg-charcoal-800 hover:bg-zinc-700 rounded-lg mb-2"
+        class="grid grid-table gap-x-0.5 mx-5 min-h-[48px] h-fit bg-charcoal-800 hover:bg-zinc-700 rounded-lg mb-2"
```
Comment (deboer-tim) [inline packages/renderer/src/lib/table/Table.svelte:168]: I'm ok with this here, but others might want this as a separate PR to allow expandable rows.

File: packages/renderer/src/lib/ingresses-routes/IngressesRoutesList.svelte [hunk 6]
```diff
@@ -0,0 +1,143 @@
+<script lang="ts">
+import { onDestroy, onMount } from 'svelte';
+
+import type { Unsubscriber } from 'svelte/store';
+import { filtered, searchPattern } from '../../stores/deployments';
```
Comment (deboer-tim) [inline packages/renderer/src/lib/ingresses-routes/IngressesRoutesList.svelte]: Needs to come from ingress + routes store - currently it is filtering deployments so I don't see the empty screen, or see it while there are still ingresses. :)

File: packages/renderer/src/lib/ingresses-routes/IngressesRoutesList.svelte [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+      }),
+    );
+    bulkDeleteInProgress = false;
+  }
+}
+
+let selectedItemsNumber: number;
+let table: Table;
+
+/* let statusColumn = new Column<IngressUI>('Status', {
```
Comment (deboer-tim) [inline packages/renderer/src/lib/ingresses-routes/IngressesRoutesList.svelte]: Needs a basic status column. Even if status is always running, needs the icon for visual design and to match other pages.

File: packages/renderer/src/lib/ingresses-routes/IngressesRoutesList.svelte [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+});
+
+let namespaceColumn = new Column<IngressUI | RouteUI, string>('Namespace', {
+  renderMapping: ingressRoute => ingressRoute.namespace,
+  renderer: SimpleColumn,
+  comparator: (a, b) => a.namespace.localeCompare(b.namespace),
+});
+
+let pathColumn = new Column<IngressUI | RouteUI, string>('Host/Path', {
+  renderer: IngressRouteColumnHostPath,
```
Comment (deboer-tim) [inline packages/renderer/src/lib/ingresses-routes/IngressesRoutesList.svelte:102]: Needs basic sorting? Would be ok if this is limited (e.g. by 'first' host) but I think it should probably be there.

File: packages/renderer/src/lib/ingresses-routes/IngressesRoutesList.svelte [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+});
+
+let backendColumn = new Column<IngressUI | RouteUI, string>('Backend', {
+  renderer: IngressRouteColumnBackend,
```
Comment (deboer-tim) [inline packages/renderer/src/lib/ingresses-routes/IngressesRoutesList.svelte:113]: Likewise, should probably have basic sorting.

File: packages/renderer/src/lib/ingresses-routes/IngressRouteColumnBackend.svelte [hunk 10]
```diff
@@ -0,0 +1,20 @@
+<script lang="ts">
+import type { IngressUI } from './IngressUI';
+import type { RouteUI } from './RouteUI';
+
+export let object: IngressUI | RouteUI;
+</script>
+
+{#if 'rules' in object && object.rules}
+  {#each object.rules as rule}
```
Comment (deboer-tim) [inline packages/renderer/src/lib/ingresses-routes/IngressRouteColumnBackend.svelte]: Maybe move the logic to ingress-route-utils.ts:getRouteUI()/getIngressUI() and stored as string[] in IngressUI/RouteUI element? Would keep the Kube logic in one place and this component simple.

File: packages/renderer/src/lib/ingresses-routes/IngressRouteColumnHostPath.svelte [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+import Link from '../ui/Link.svelte';
+import type { IngressUI } from './IngressUI';
+import type { RouteUI } from './RouteUI';
+
+export let object: IngressUI | RouteUI;
+</script>
+
+{#if 'rules' in object && object.rules}
+  {#each object.rules as rule}
+    {@const host = rule.host ?? ''}
```
Comment (deboer-tim) [inline packages/renderer/src/lib/ingresses-routes/IngressRouteColumnHostPath.svelte]: Ditto to Backend comment, maybe move logic to util class and store array in IngressUI/RouteUI, keeps UI component simpler and just about display.

**Review by deboer-tim — COMMENTED** [review]
Needs tests (can copy a lot from deployments, elsewhere), and added a bunch of general comments (most of which you probably already knew). It worked perfectly fine for me though 👍🏼, once you're done these I think it's ready for review.

**Comment by deboer-tim** [comment]
For completeness you should do the same thing (updates + tests) as #5370 too.

Since this may come up for review while I'm still on PTO, just adding a comment to say there is no reason to wait for me - if the changes above are made I would expect to be +1ing this and anyone else can review.

File: packages/renderer/src/lib/ingresses-routes/ingress-route-utils.ts [hunk 12]
```diff
@@ -21,11 +21,17 @@ import type { IngressUI } from './IngressUI';
...
+  label: string;
+  url?: string;
+}
+
 export class IngressRouteUtils {
   getIngressUI(ingress: V1Ingress): IngressUI {
     return {
       name: ingress.metadata?.name || '',
       namespace: ingress.metadata?.namespace || '',
+      status: 'RUNNING',
```
Comment (axel7083) [inline packages/renderer/src/lib/ingresses-routes/ingress-route-utils.ts:34]: @deboer-tim The discussion https://github.com/containers/podman-desktop/issues/4404 is closed but I was wondering the purpose of the `status` field ? 
it seems to always be `RUNNING` and __I don't think__ it is possible to have other status for an ingress ?

> In the mockup, they seems to be able to be __disabled__ but I don't see how, maybe you are referring to the transition state, `creating`, `deleting` but they are probably going to be instantaneous compared to pods having a lot of states

File: packages/renderer/src/lib/ingresses-routes/IngressRouteColumnHostPath.svelte [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+import type { RouteUI } from './RouteUI';
+import { IngressRouteUtils } from './ingress-route-utils';
+
+export let object: IngressUI | RouteUI;
+
+const ingressRouteUtils = new IngressRouteUtils();
+</script>
+
+{#each ingressRouteUtils.getHostPaths(object) as hostPath}
+  <div class="text-sm text-gray-500">
```
Comment (axel7083) [inline packages/renderer/src/lib/ingresses-routes/IngressRouteColumnHostPath.svelte]:
```suggestion
  <div class="text-sm text-gray-500 overflow-hidden text-ellipsis">
```

| Before | After |
| --- | --- |
| image | image |

**Review by axel7083 — COMMENTED** [review]
Amazing work! I'm adding a screenshot to show the wonderful work you made!

image

File: packages/renderer/src/lib/ingresses-routes/ingress-route-utils.ts (hunk 12 above)
Comment (deboer-tim) [reply packages/renderer/src/lib/ingresses-routes/ingress-route-utils.ts:34]: Yeah, an ingress isn't RUNNING like other objects, but using it allows us to use the status icon, be consistent with other pages, and states like 'deleting' will show up similarly (if there is any delay). Not totally 'correct', but I think there are benefits until/unless we come up with another way to do this.

**Review by deboer-tim — APPROVED** [review]
Reviewed all the code and tests again, and LGTM. I'd pick up Axel's suggestion if you have a chance, but I can always merge that later.

File: packages/renderer/src/lib/ingresses-routes/IngressRouteColumnHostPath.svelte (hunk 13 above)
Comment (lstocchi) [reply packages/renderer/src/lib/ingresses-routes/IngressRouteColumnHostPath.svelte]: Thanks @axel7083 👍

File: packages/renderer/src/lib/ingresses-routes/IngressRouteActions.spec.ts [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+const deleteRoutesMock = vi.fn();
+
+beforeEach(() => {
+  (window as any).kubernetesDeleteIngress = deleteIngressMock;
+  (window as any).kubernetesDeleteRoute = deleteRoutesMock;
+});
+
+afterEach(() => {
+  vi.resetAllMocks();
+  vi.clearAllMocks();
```
Comment (benoitf) [inline packages/renderer/src/lib/ingresses-routes/IngressRouteActions.spec.ts:37]: reported by @dgolovin one day, reset is doing the clear so we can only do the reset (but I know we have a lot of places with these 2lines, not a requested change)

File: packages/renderer/src/lib/ingresses-routes/IngressesRoutesList.spec.ts [hunk 15]
```diff
@@ -0,0 +1,183 @@
+/**********************************************************************
+ * Copyright (C) 2023 Red Hat, Inc.
```
Comment (benoitf) [inline packages/renderer/src/lib/ingresses-routes/IngressesRoutesList.spec.ts]:
```suggestion
 * Copyright (C) 2024 Red Hat, Inc.
```

**Review by benoitf — APPROVED** [review]

**Pull request merged**